File: keemsebin/Implementation/BOJ_22856.js
const input = require("fs").readFileSync("/dev/stdin").toString().split("\n");

const N = Number(input[0]);
const tree = [];
let count = 0;
let lastVisited = 0;
for (let i = 0; i < N; i++) {
  const [node, left, right] = input[i + 1].split(" ").map(Number);
  tree[node] = [left, right];
}

function similarInOrder(node) {
  if (node === -1) return;

  similarInOrder(tree[node][0]);
  if (lastVisited !== 0) count++;
  lastVisited = node;
  similarInOrder(tree[node][1]);
}

similarInOrder(1);
console.log(count);
// [1, 2, 4, 4, 2, 5, 5, 1, 3, 6, 6, 3, 7, 7]

File: comgongdaeng/Implementation/BOJ_22856.py
# 이진 트리 -> 배열? 트리?
# 왼쪽 자식이 있다면 방문 / 오른쪽 자식이 있다면 방문 / # 부모가 있다면 방문 +1
# 마지막 노드인 경우 +1 하지 않음
import sys
sys.setrecursionlimit(10**7)
n = int(sys.stdin.readline())
tree=  [[] for _ in range(n+1)]

for i in range(n):
    root, left, right = list(map(int, sys.stdin.readline().split()))
    if left == -1: left = 0
    if right == -1: right = 0
    tree[root].append(left)
    tree[root].append(right)

inOrder = []
def in_order(node):
    if tree[node][0]:
        in_order(tree[node][0])
    inOrder.append(node)
    if tree[node][1]:
        in_order(tree[node][1])

def sim_in_order(node):
    global cnt
    global ans
    if node == inOrder[-1]:
        ans = cnt
    if tree[node][0]:
        cnt += 1
        sim_in_order(tree[node][0])
        cnt += 1
        if node == inOrder[-1]:
            ans = cnt
    if tree[node][1]:
        cnt += 1
        sim_in_order(tree[node][1])
        cnt += 1

cnt = 0
in_order(1)
ans = 0
sim_in_order(1)
print(ans)